pages/blog/2020-03-12-nextjs-without-client-side-react.md

```diff
@@ -0,0 +1,164 @@
+
+The other option I did consider was the plain vanilla ice-cream flavour. Simple, no build step, no Markdown, only plain HTML, JS and CSS files. This was tempting. I usually opt for this if the project allows it, but as always, as soon as you want to do ANYTHING fancy like component re-use, that approach quickly get's messy and hard to maintain.
+
+## What's _Next?_
+
+<Disclaimer>Disclaimer: you can skip this part if you know what Next is.</Disclaimer>
+
+Next is a multi-purpose React application framework, but for my purposes it's mainly a static site generation tool (could be compare to Jekyll or Gatsby). It takes an input (Markdown in my case) and spits out a bunch of HTML files that are later deployed on a CDN. No server-side runtime required.
+
+### Next is great for universal apps
+
+Out of the box Next apps are isomorphic/universal. That means that most of the JavaScript that you write is automatically bundled, tree-shaken, split into small modules and shipped to the client. This allows you to server side render your pages, and add interactivity using React on the client. This is all very clever, and super cool if you need it. I don't.
+
+### The Elephant In the Room: 453kB of JavaScript
+
+So what's the problem? Ladies and Gentlemen, this is a blog. It's a one way communication channel with little or no interactivity. Most people read one article and then leave, and that's OK.
+
+For the tiny amount of interactivity that this site offers (more on this later) there is simply no way I could justify sending you 453kb of JavaScript, that adds no value.
+
+
+![Client Side Bundle](/client-side-bundle.jpg "What are you saying? All of this JavaScript? For no value!?!")
+
+
+## Part 2: Killing the Elephant
+
+What I like about Next is that you can heavily customize it. I've done so to an extent that there's no notion of the Next runtime or traces of it in the statically generated HTML.
+
+What I've done here is to remove all client-side JavaScript and extra markup that Next.js outputs. This is how:
+
+### Create Custom head component
+
+One option is to just use a normal HTML `<head>` element here. However, extending the Next Head component will allows us to inject a `<title>` from any component or page, this is useful.
+
+```jsx
+import { Head } from "next/document";
+class MyHead extends Head {
+  render() {
+    let { head } = this.context._documentProps;
+    let children = this.props.children;
+    return (
+      <head {...this.props}>
+        {children}
+        {head}
+      </head>
+    );
+  }
+}
+```
+
+### Custom Main component
+
+This is not really necceary but to satisfy my mild OCD I wanted to get rid of the `<div id="next">` wrapper around my content.
+
+```jsx
+import { Main } from "next/document";
```

Comment:
I will add comments to the code explaining what is happening given this is next.js specific.